### DISCUSSION TITLE
What exactly does "Incompatible schema cookies. Please, restart from old backup." on joining node mean?

### DISCUSSION THREAD

**Discussion by flbraun** [emqx/emqx#13629] [Q&A]
I have an EMQX 5.7.0 cluster with one core node and some replicas. The cluster has been up for about 2 months and is seemingly operational, accepts connections, delivers messages, etc.

When I try to add another core node to the cluster, the new node fails to join the cluster and gives the following log message:
```
Generic server ekka_autocluster terminating. Reason: {bad_return_value,{error,{merge_schema_failed,\"Incompatible schema cookies. Please, restart from old backup.'emqx@<ip>' = <redacted>
```

From my very limited knowledge of EMQX's internals, Erlang and its ecosystem I understand that the new node starts a new mnesia node and attempts to sync with another member of the cluster, but that fails because the new and existing mnesia nodes do not have the same schema, although I would have expected the new node to simply adopt the schema of the existing node.

Unfortunately I can't apply a backup as suggested by the error message as I don't have any (the cluster is in a non-critical experimental environment), but instead of nuking and rebuilding the entire cluster I would like to better understand what's happening and if/how I could rescue it.

**Comment by zmstone** [comment]
Hi @flbraun 
Could you provide your clustering configs and the steps how you add a new core node to the cluster ?
I'll try to see if I can reproduce it first.

**Comment by flbraun** [comment]
@zmstone Thanks for reaching out!

The cluster is deployed via AWS ECS. The task definition specifies the following environment variables:
```json
"environment": [
    {
        "name": "EMQX_CLUSTER__AUTOCLEAN",
        "value": "5m"
    },
    {
        "name": "EMQX_CLUSTER__AUTOHEAL",
        "value": "true"
    },
    {
        "name": "EMQX_CLUSTER__DISCOVERY_STRATEGY",
        "value": "dns"
    },
    {
        "name": "EMQX_CLUSTER__DNS__RECORD_TYPE",
        "value": "a"
    },
    {
        "name": "EMQX_CLUSTER__DNS__NAME",
        "value": "mycluster.example"
    },
    {
        "name": "EMQX_NODE__DB_BACKEND",
        "value": "rlog"
    },
    {
        "name": "EMQX_NODE__DB_ROLE",
        "value": "core"
    },
    {
        "name": "EMQX_NODE__MAX_PORTS",
        "value": "2097152"
    },
    {
        "name": "EMQX_NODE__PROCESS_LIMIT",
        "value": "2097152"
    }
],
```

I have another two clusters running the exact same task definition and these are behaving well. Am curious to see if you can reproduce it.

**Comment by zmstone** [comment]
Hi @flbraun 
Sorry for the delay.
It does not sound like you have a exotic deployment, but rather standard.

The log is truncated, could you provide the log in full details?
Because it should include which exact table caused the incompatibility.

**Comment by zmstone** [comment]
Hi again @flbraun 
Could you check if the node has an unclean `data/mnesia` directory before it joins the cluster? 
Since you mentioned it's a 'new' node, I would expect the data directory is empty, but does not contain data from an old run.
